hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystem.java

```diff
@@ -410,61 +416,183 @@ public boolean rename(final Path src, final Path dst) throws IOException {
+        if (makeQualified(nestedDstParent).equals(qualifiedSrcPath)) {
+          //testRenameChildDirForbidden.
+          LOG.info("Rename src: {} dst: {} failed as dst is subDir of src",
+              qualifiedSrcPath, qualifiedDstPath);
+          return false;
+        }
+        nestedDstParent = nestedDstParent.getParent();
+      }
+    }
+
+
     TracingContext tracingContext = new TracingContext(clientCorrelationId,
         fileSystemId, FSOperationType.RENAME, true, tracingHeaderFormat,
         listener);
     // rename under same folder;
-    if(makeQualified(parentFolder).equals(qualifiedDstPath)) {
+    if (makeQualified(parentFolder).equals(qualifiedDstPath)) {
       return tryGetFileStatus(qualifiedSrcPath, tracingContext) != null;
     }
 
-    FileStatus dstFileStatus = null;
     if (qualifiedSrcPath.equals(qualifiedDstPath)) {
       // rename to itself
       // - if it doesn't exist, return false
       // - if it is file, return true
       // - if it is dir, return false.
-      dstFileStatus = tryGetFileStatus(qualifiedDstPath, tracingContext);
-      if (dstFileStatus == null) {
+      final AtomicBoolean isDstDirectory = new AtomicBoolean();
+      final AtomicBoolean isDstExists = new AtomicBoolean();
+
+      getPathInformation(qualifiedDstPath, tracingContext, isDstDirectory,
+          isDstExists);
+      if (!isDstExists.get()) {
         return false;
       }
-      return dstFileStatus.isDirectory() ? false : true;
+      return isDstDirectory.get() ? false : true;
     }
 
+    final AtomicBoolean isDstDirectory = new AtomicBoolean();
+    final AtomicBoolean isDstExists = new AtomicBoolean();
+
     // Non-HNS account need to check dst status on driver side.
-    if (!abfsStore.getIsNamespaceEnabled(tracingContext) && dstFileStatus == null) {
-      dstFileStatus = tryGetFileStatus(qualifiedDstPath, tracingContext);
+    if (!abfsStore.getIsNamespaceEnabled(tracingContext)) {
+      getPathInformation(qualifiedDstPath, tracingContext, isDstDirectory,
+          isDstExists);
     }
 
     try {
       String sourceFileName = src.getName();
       Path adjustedDst = dst;
 
-      if (dstFileStatus != null) {
-        if (!dstFileStatus.isDirectory()) {
+      if (isDstExists.get()) {
+        if (!isDstDirectory.get()) {
           return qualifiedSrcPath.equals(qualifiedDstPath);
         }
         adjustedDst = new Path(dst, sourceFileName);
+        qualifiedDstPath = makeQualified(adjustedDst);
+        LOG.debug("Qualified dst path: {}", qualifiedDstPath);
+        /*
+         * For blob-endpoint with nonHNS account, check if the qualifiedDstPath
+         * exist in backend. If yes, HTTP_CONFLICT exception has to be thrown.
+         */
+        if (getAbfsStore().getAbfsConfiguration().getPrefixMode()
+            == PrefixMode.BLOB) {
+          isDstDirectory.set(false);
+          isDstExists.set(false);
+          getPathInformation(qualifiedDstPath, tracingContext, isDstDirectory,
```

Comment:
This has already been computed in the previous if check where namespace is not enabled.

Reply:
No. This is different. This happens because src is a dir and dst is a dir, the new qualifiedDstPath shall be 
/dst/src.